### PR TITLE
Remove reflection-based deserialization setters

### LINKED ISSUE #16353: Repository Quality: Reflection Anti-patterns and Private-Setter Bypass Audit
### 🎯 Repository Quality Improvement Report — Reflection Anti-patterns and Private-Setter Bypass Audit

**Analysis Date**: 2026-08-12
**Focus Area**: reflection-anti-patterns-private-setter-bypass
**Strategy Type**: Custom

### Executive Summary

An audit of the 46 source files that use reflection reveals a cluster of anti-patterns centred on **bypassing private setters during deserialization** and **late-binding to external assemblies via string-based method lookup**. The most critical findings:

- `TestSessionInfo.Id` has `private set`; two separate converters use `GetProperty(...).SetValue(...)` reflection to bypass it — incompatible with NativeAOT/trimming.
- `DiscoveryCriteria` has four `private set` properties bypassed in both the legacy Jsonite path and System.Text.Json path via reflection. The Jsonite fallback also uses the compiler-internal `<{prop.Name}>k__BackingField` naming convention.
- `CodeCoverageDataAttachmentsHandler.cs` retrieves `Task.Result` via `GetProperty("Result").GetValue(...)` when a direct cast to `Task<IList<string>>` is sufficient.

<details>
<summary><b>Full Analysis</b></summary>

### Metrics

| Metric | Value | Status |
|--------|-------|--------|
| Source files using reflection APIs | 46 | ⚠️ |
| `Activator.CreateInstance` call sites | 16 | ⚠️ |
| `Type.GetType` call sites | 6 | ⚠️ |
| Private-setter bypasses via `SetValue` | 6 across 4 converters | ❌ |
| Compiler-internal `k__BackingField` hacks | 1 | ❌ |
| `Task.Result` accessed via reflection | 1 | ❌ |

### Key Findings

**[Critical] TestSessionInfo.Id private-setter bypass — duplicated across two converters**
`JsoniteConvert.cs:409` and `TestSessionInfoConverter.cs:36` both execute:
```csharp
typeof(TestSessionInfo).GetProperty(nameof(TestSessionInfo.Id))!.SetValue(info, id);
```
Root cause: `TestSessionInfo` has no constructor that accepts a known `Guid`.

**[Critical] DiscoveryCriteria private-setter bypass in two serialization paths**
`JsoniteConvert.cs:417-420` and `DiscoveryCriteriaConverter.cs:45-48` both use `GetProperty(...).SetValue(...)` for `AdapterSourceMap`, `FrequencyOfDiscoveredTestsEvent`, `DiscoveredTestEventTimeout`, and `RunSettings`. The Jsonite path also falls back to writing directly into `<{prop.Name}>k__BackingField` — an undocumented compiler implementation detail.

**[High] Task.Result via reflection in CodeCoverageDataAttachmentsHandler.cs**
`CodeCoverageDataAttachmentsHandler.cs:135`: `task.GetType().GetProperty("Result")!.GetValue(task, null)` — a direct cast to `Task<IList<string>>` eliminates the reflection call.

**[Medium] Null-safety gap in TestExtensions.cs**
`Type.GetType(testPluginInformation?.AssemblyQualifiedName!)` — the `!` suppresses the warning but `AssemblyQualifiedName` can still be null at runtime; `Type.GetType(null)` returns null silently.

</details>

---

#### Task 1: Add `internal` deserialization constructor to `TestSessionInfo`

**Priority**: High | **Effort**: Small

Add `internal TestSessionInfo(Guid id) { Id = id; }` to `src/Microsoft.TestPlatform.ObjectModel/Client/TestSessionInfo.cs`, then replace `GetProperty(...).SetValue(...)` in both:
- `src/Microsoft.TestPlatform.CommunicationUtilities/Serialization/JsoniteConvert.cs:408-409`
- `src/Microsoft.TestPlatform.CommunicationUtilities/Serialization/TestSessionInfoConverter.cs:35-36`

---

#### Task 2: Expose `internal set` on `DiscoveryCriteria` private-setter properties

**Priority**: High | **Effort**: Medium

Change `private set` to `internal set` on `AdapterSourceMap`, `FrequencyOfDiscoveredTestsEvent`, `DiscoveredTestEventTimeout`, and `RunSettings` in `src/Microsoft.TestPlatform.ObjectModel/Client/DiscoveryCriteria.cs`. Update both converters to assign directly rather than via `GetProperty(...).SetValue(...)`.

---

#### Task 3: Remove the `k__BackingField` compiler-internal hack from `JsoniteConvert.cs`

**Priority**: Medium | **Effort**: Small

After Tasks 1 and 2 land, verify no other serialized type depends on the backing-field fallback (`JsoniteConvert.cs:563`) and remove it. Replace with an `EqtTrace.Warning` when no accessible setter exists, making failures visible.

---

#### Task 4: Replace `Task.Result` reflection in `CodeCoverageDataAttachmentsHandler.cs`

**Priority**: Medium | **Effort**: Small

Replace `task.GetType().GetProperty("Result")!.GetValue(task, null)` (`CodeCoverageDataAttachmentsHandler.cs:135`) with a direct cast to `Task<IList<string>>`. Also add a check that `s_mergeMethodInfo.ReturnType` is assignable to `Task<IList<string>>` when caching the `MethodInfo` to fail fast at initialization rather than at call time.

---

#### Task 5: Fix null-safety gap in `TestExtensions.cs` and document `throwOnError` in `TestPluginManager.cs`

**Priority**: Low | **Effort**: Small

- `TestExtensions.cs:431`: replace `testPluginInformation?.AssemblyQualifiedName!` with an explicit null guard before passing to `Type.GetType`.
- `TestPluginManager.cs:36`: wrap `Type.GetType(extensionTypeName, true)` in a targeted `catch (TypeLoadException)` that logs the type name at error level, so the failure is not swallowed by the broad `catch (Exception)` in callers.

---

### 📊 Historical Context

<details>
<summary><b>Previous Focus Areas (last 10)</b></summary>

| Date | Focus Area | Type |
|------|------------|------|
| 2026-08-11 | conditional-compilation-dead-code-and-legacy-symbol-hygiene | Custom |
| 2026-08-10 | permanently-ignored-test-coverage-debt | Custom |
| 2026-08-07 | timeout-budget-architecture-and-consistency | Custom |
| 2026-08-06 | assembly-architecture-inference-coverage | Custom |
| 2026-08-05 | environment-variable-propagation-testhost-setup | Custom |
| 2026-08-04 | test-case-filter-engine-correctness-and-edge-cases | Custom |
| 2026-08-03 | telemetry-opt-in-correctness-and-data-hygiene | Custom |
| 2026-07-31 | testproperty-propertybag-type-safety-and-serialization-correctness | Custom |
| 2026-07-30 | logger-concurrency-correctness-and-output-reliability | Custom |
| 2026-07-29 | ipc-connection-resilience-timeout-arithmetic-correctness | Custom |

</details>

*Next analysis: 2026-08-13 — Focus area selected based on diversity algorithm*

> Generated by [Repository Quality Improver](https://github.com/microsoft/vstest/actions/runs/31562886354) · 62.6 AIC · ⌖ 8.48 AIC · ⊞ 8.9K · [◷](https://github.com/search?q=repo%3Amicrosoft%2Fvstest+is%3Aissue+%22gh-aw-workflow-call-id%3A+microsoft%2Fvstest%2Frepository-quality-improver%22&type=issues)
> - [x] expires  on Aug 14, 2026, 4:26 AM UTC

### PR DESCRIPTION
**Pull request by nohwnd**
Reflection was used to write private ObjectModel setters in both serializers, and Jsonite also depended on compiler-generated backing-field names.

This gives CommunicationUtilities internal access to the deserialization state and assigns it directly. Jsonite reconstructs immutable data-collection IDs through their constructors instead of writing backing fields. The optional code coverage loader still uses reflection, but awaits its declared task result directly, and a test pins that declared return type so a package update that changes it fails instead of silently returning unmerged files. Invalid plugin metadata is now diagnosed and skipped.

The NativeAOT consumer that guards this round trips through the public `JsonDataSerializer` entry point, so it needs no friend access and no signing.

Fix #16353

Verified: `build.cmd -c Release`; CommunicationUtilities 614/614 and Utilities 124/124 on net11.0 and net481, including the NativeAOT publish and runtime round trip.

🤖

### CODE REVIEW

File: test/Microsoft.TestPlatform.CommunicationUtilities.UnitTests/NativeAotCompatibilityTests.cs [hunk 1]
```diff
@@ -51,14 +51,23 @@ public void TranslationLayer_NativeAotPublish_ShouldNotProduceSerializationWarni
...
         };
 
+        if (OperatingSystem.IsWindows())
+        {
+            var visualStudioInstallerPath = Path.Combine(
+                Environment.GetFolderPath(Environment.SpecialFolder.ProgramFilesX86),
+                "Microsoft Visual Studio",
+                "Installer");
+            psi.Environment["PATH"] = $"{visualStudioInstallerPath}{Path.PathSeparator}{psi.Environment["PATH"]}";
+        }
```
[the remark this answers is not in the document]
Comment (nohwnd) [reply test/Microsoft.TestPlatform.CommunicationUtilities.UnitTests/NativeAotCompatibilityTests.cs:72]: Good catch, fixed in 462bafbe8f. `psi.Environment` is read with `TryGetValue` now, and when `PATH` is missing it is set to the Visual Studio Installer directory on its own instead of throwing.

🤖

> [!NOTE]
> Auto-resolved by the [GitHub Copilot app](https://github.com/github/app).

**Comment by nohwnd** [comment]
The duplicate #16375 picked up one extra reviewer comment on the same line, suggesting a fallback to `Environment.GetEnvironmentVariable("PATH")` when `psi.Environment` has no `PATH`. That reads the same source, so it returns nothing in exactly the case it is meant to cover, and the dictionary is already OrdinalIgnoreCase on Windows. Answered and resolved there: https://github.com/microsoft/vstest/pull/16375#discussion_r3801386539

No change needed here.

🤖

**Comment by nohwnd** [comment]
Everything here is in main already. #16356 merged the same work, so 16 of the 17 files match main exactly and this branch now conflicts.

The one piece that got lost is the `PATH` guard from 462bafbe8f. #16356 merged the version that still reads `psi.Environment["PATH"]` through the indexer, so the fix never reached main. It is now in #16380 on its own.

On the suppressed comment about `DiscoveryCriteriaFactory.Create` and `frequencyOfDiscoveredTestsEvent`: the converter always writes that property, and every sender builds `DiscoveryCriteria` through the same validating constructor, so a payload with a missing or zero frequency can only be hand-written or corrupted. Clamping it would mean inventing a batch size and hiding a malformed message, so I would rather it throw.

Closing this as superseded.

🤖
